nextstrain/cli/remote/nextstrain_dot_org.py

```diff
@@ -0,0 +1,656 @@
+
+            if not narratives_only(path):
+                narrative = f"narratives/{narrative}"
+
+            endpoint = destination = api_endpoint(path if single else path / narrative)
+
+            yield file, destination
+
+            put(endpoint, file, markdown_type)
+
+
+def download(url: urllib.parse.ParseResult, local_path: Path, recursively: bool = False) -> Iterable[Tuple[str, Path]]:
+    """
+    Download the datasets or narratives deployed at the given remote *url*,
+    optionally *recursively*, saving them into the *local_dir*.
+    """
+    path = remote_path(url)
+
+    if str(path) == "/" and not recursively:
+        raise UserError(f"""
+            No path specified in URL ({url.geturl()}); nothing to download.
+
+            Did you mean to use --recursively?
+            """)
+
+    authorization = auth_headers()
+
+    with requests.Session() as http:
+        resources = _ls(path, recursively = recursively, http = http, authorization = authorization)
+
+        if not resources:
+            raise UserError(f"Path {path} does not seem to exist")
+
+        for resource in resources:
+            for subresource in resource.subresources:
+                # Remote source
+                endpoint = source = api_endpoint(resource.path)
+
+                if not subresource.primary:
+                    source += f" ({sidecar_suffix(subresource.media_type)})"
+
+                response = http.get(
+                    endpoint,
+                    headers = {"Accept": subresource.media_type, **authorization},
+                    stream = True)
+
+                with response:
+                    # Skip/ignore missing sidecars
+                    if response.status_code == 404 and not subresource.primary:
+                        continue
+
+                    # Check for bad response
+                    raise_for_status(response)
+                    assert content_media_type(response) == subresource.media_type
+
+                    # Local destination
+                    if local_path.is_dir():
+                        if recursively:
+                            # If we're recursively downloading under a path
+                            # which is itself a dataset that exists, e.g.
+                            # `nextstrain remote download -r a/b` when the
+                            # following datasets exist:
+                            #
+                            #       a/b
+                            #       a/b/x
+                            #       a/b/y
+                            #       a/b/z
+                            #
+                            # then we need to include a little of the given
+                            # path in the local filenames, e.g. strip only "a",
+                            # leaving "b".  Otherwise, e.g. if a/b doesn't
+                            # exist itself only a/b/x, etc, then we can strip
+                            # "a/b" entirely.
```

Comment:
Hmm, this behavior was very unexpected for me when I was testing the recursive download.

Flu datasets might be a very specific use case:
1. run `remote download -r groups/test/flu /tmp/testing` 
2. run `remote download -r groups/test/flu/seasonal /tmp/testing` 

I expected the second command to just overwrite the existing files within `/tmp/testing`. Instead I got two sets of the same files but with slightly different names because the prefixes were stripped.

Reply:
Well, "very unexpected" is not what we want! :joy: It's a good point that it's maybe a bit weird that downloading the same datasets two different ways produces different local filenames. I wonder if that situation is likely to crop up in typical usage as opposed to testing? I'd liked keeping filenames shorter than longer (e.g. instead of using the whole URL path) and picking a place to shorten them at that was under the user's control rather than picked/hardcoded by us. But maybe it's too confusing.

There's lots of brainstorming in my notes about possible behaviours here, and there were one or two decent discussions in Slack a while back, but nothing very conclusive. Maybe I should go back and summarize it all in writing and bring it back for wider consideration as part of (or separate from) this PR? Definitely would appreciate more input here.

Reply:
(Ah, I just realized that I completely missed the flu example in the [download help docs](https://github.com/nextstrain/cli/pull/141/commits/a197ac4c87e63c6505e8d5785188234427f1f7aa#diff-c5bff80b5b1c30b6824da28ca7b0d22d0f9d41c8adb2c8b4b94833d664ed0bbf))

I actually prefer the opposite: keep the whole URL path so that the user can see exactly what they downloaded without having to inspect the file. They can easily rename the files locally. 

Another problem is overwriting  since the path prefixes are used to scope/filter downloads. 
If I want to download H3N2 and H1N1pdm builds to the same directory and I run
1. download nextstrain.org/flu/seasonal/h3n2/ 
2. download nextstrain.org/flu/seasonal/h1n1pdm/

The second command would just overwrite my first download 😞

Reply:
(Relevant [Slack discussion](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1628543130011300) from a while back)

Reply:
> I actually prefer the opposite: keep the whole URL path so that the user can see exactly what they downloaded without having to inspect the file.

I see the appeal of keeping the whole URL (though think doing `/` → `_` transformation on the whole URL is weird), but I guess I don't get why someone wouldn't know what they just downloaded.

> They can easily rename the files locally.

For a few files, yes. I'm not so confident in this being a reasonable expectation if someone does `nextstrain remote download -r nextstrain.org/ncov` (or worse, for flu) and ends up with 42 files (2 databases (GISAID, Open) × 7 regions × 3 files (main, root-sequence, tip-frequencies)).

One of the reasons for `download` using the shortest local filenames that are still distinguishing is that it helps play nice with `upload` when handling multiple datasets. Although `upload` doesn't care about local filename if uploading a single dataset, the filename still matters for multiple uploads and becomes a suffix on the given destination URL.

I'm imagining that a download + upload combo would be used for copying files between Groups, but maybe that would be less common than "download, do secondary analyses, never upload back", for which you'd tend to want longer more descriptive filenames. (The primary usage is still "run a build, upload files", and `download` doesn't matter for that.)

Reply:
> Another problem is overwriting since the path prefixes are used to scope/filter downloads. If I want to download H3N2 and H1N1pdm builds to the same directory and I run
> 
>     1. download nextstrain.org/flu/seasonal/h3n2/
> 
>     2. download nextstrain.org/flu/seasonal/h1n1pdm/
> 
> 
> The second command would just overwrite my first download

Yeah, this is not ideal. The workaround is downloading each into a specific directory (or, in the nearish future, giving a filename prefix instead of a dir).

This comes back to the Slack discussion from the fall, where you pointed out that providing a toggle between two different ways of naming local files may be best. What is better for the "download then re-upload" workflow is not better for the "download and analyze" workflow and vice versa.